cpu/avr8_common/include/cpu.h

```diff
@@ -183,8 +182,8 @@ static inline void __attribute__((always_inline)) cpu_print_last_instruction(voi
         : /* no inputs */
         : /* no clobbers */
     );
-    ptr = hi << 8 | lo;
-    printf("Stack Pointer: 0x%04x\n", ptr);
+
+    return hi << 8 | lo;
```

Comment:
Well, this is the stack pointer, not the last instruction.

With this, it *almost* works:
```diff
diff --git a/cpu/avr8_common/avr8_cpu.c b/cpu/avr8_common/avr8_cpu.c
index 2323236e97..b5b0926067 100644
--- a/cpu/avr8_common/avr8_cpu.c
+++ b/cpu/avr8_common/avr8_cpu.c
@@ -72,6 +72,33 @@ uint8_t avr8_state = 0;
 
 void get_mcusr(void) __attribute__((naked, section(".init0"), used));
 
+uintptr_t cpu_get_last_instruction(void)
+{
+    /* There is absolutely no way to read out the program counter of an AVR.
+     * However, it will be pushed to the stack so that a ret instruction can
+     * pop it and jump to it. We just pop the address here, and push it back to
+     * the stack. Note that this relies on clever register allocation of the
+     * compiler, so that `result` is directly allocated in the registers that
+     * according to the calling convention contain the result. This does work
+     * for -Os, but with -O0 or -Og it will likely fail
+     */
+    uintptr_t result;
+    __asm__ volatile(
+        "pop %B[dest]"              "\n\t"
+        "pop %A[dest]"              "\n\t"
+        "push %A[dest]"             "\n\t"
+        "push %B[dest]"             "\n\t"
+        : [dest]    "=r"(result)
+        : /* no inputs */
+        : /* no clobbers */
+    );
+
+    /* program counter is given as (16 bit) word address (instructions on AVR
+     * are always 16 bit wide), rather than as byte address. Let's convert it
+     * here to make life for people using avr-addr2line easier */
+    return result << 1;
+}
+
 void get_mcusr(void)
 {
     /* save soft reset flag set in reset routine */
diff --git a/cpu/avr8_common/include/cpu.h b/cpu/avr8_common/include/cpu.h
index c3eb747c37..7329a4a1e2 100644
--- a/cpu/avr8_common/include/cpu.h
+++ b/cpu/avr8_common/include/cpu.h
@@ -169,22 +169,7 @@ void avr8_clk_init(void);
 /**
  * @brief   Get the last instruction's address
  */
-static inline uintptr_t __attribute__((always_inline)) cpu_get_last_instruction(void)
-{
-    uint8_t hi;
-    uint8_t lo;
-
-    __asm__ volatile (
-        "in %[hi], __SP_H__     \n\t"
-        "in %[lo], __SP_L__     \n\t"
-        : [hi] "=r"(hi),
-          [lo] "=r"(lo)
-        : /* no inputs */
-        : /* no clobbers */
-    );
-
-    return hi << 8 | lo;
-}
+uintptr_t cpu_get_last_instruction(void);
 
 /**
  * @brief   Initializes avrlibc stdio
```

```
2022-06-17 19:47:28,813 # main(): This is RIOT! (Version: 2022.07-devel-797-g54c0f0-ben)
2022-06-17 19:47:28,826 # Hello World!
2022-06-17 19:47:28,833 # 0xef0
2022-06-17 19:47:28,857 # *** RIOT kernel panic:
2022-06-17 19:47:28,875 # FAILED ASSERTION.
2022-06-17 19:47:28,876 # 
2022-06-17 19:47:28,889 # *** halted.

$ avr-addr2line -e bin/m1284p/hello-world.elf 0xef0
/home/maribu/Repos/software/RIOT/core/lib/assert.c:30
```

Couldn't we have a `uintptr_t cpu_get_program_counter(void)` instead? The `assert()` macro would pass the return value of that to `_assert_panic()` as argument?

Reply:
Isn't that unrelated to this PR? I.e., was apparently broken before.

Reply:
> Couldn't we have a uintptr_t cpu_get_program_counter(void) instead? The assert() macro would pass the return value of that to _assert_panic() as argument?


The assert marco cannot call any function of cpu.h since assert.h may not provide other headers content (only _functions)

One should not do a conversion (bitshift of the pc-value stored on the avr stack) without extending the Typ to at least 22-bit (or 17Bit) (of cause uint32_t would match)

Reply:
One should not do a conversion (bitshift of the pc-value stored on the avr stack) without extending the Typ to at least 22-bit (or 17Bit) (of cause uint32_t would match) - the avr really use all 16 bits of their uintptr_t

Reply:
for the clever bit of the compiler wouldn't the `register` memory work in any case (`register uint8_t hi;`)

Reply:
With some limitations (see details part in the documentation) this actually works:

```diff
diff --git a/cpu/avr8_common/include/cpu.h b/cpu/avr8_common/include/cpu.h
index c3eb747c37..4e230333f1 100644
--- a/cpu/avr8_common/include/cpu.h
+++ b/cpu/avr8_common/include/cpu.h
@@ -168,22 +168,30 @@ void avr8_clk_init(void);
 
 /**
  * @brief   Get the last instruction's address
+ *
+ * @details There is absolutely no way to do this reliably on AVR. Instead, we
+ *          examine the stack frame and assume that the caller called this
+ *          function right away, so that the return address is still at the top
+ *          of the stack
  */
-static inline uintptr_t __attribute__((always_inline)) cpu_get_last_instruction(void)
+__attribute__((always_inline))
+static inline uintptr_t cpu_get_last_instruction(void)
 {
-    uint8_t hi;
-    uint8_t lo;
-
-    __asm__ volatile (
-        "in %[hi], __SP_H__     \n\t"
-        "in %[lo], __SP_L__     \n\t"
-        : [hi] "=r"(hi),
-          [lo] "=r"(lo)
+    uintptr_t result;
+    __asm__ volatile(
+        "pop %B[dest]"              "\n\t"
+        "pop %A[dest]"              "\n\t"
+        "push %A[dest]"             "\n\t"
+        "push %B[dest]"             "\n\t"
+        : [dest]    "=r"(result)
         : /* no inputs */
-        : /* no clobbers */
+        : "memory"
     );
 
-    return hi << 8 | lo;
+    /* Result now contains instruction to return to, subtract one to get
+     * the instruction that called this function. Also multiply by two to get
+     * the byte position, rather than the (16 bit) instruction position */
+    return (result - 1) * 2;
 }
 
 /**
```

```
2022-06-17 21:16:41,443 # main(): This is RIOT! (Version: 2022.07-devel-797-g54c0f0-ben)
2022-06-17 21:16:41,456 # Hello World!
2022-06-17 21:16:41,462 # 0x104
2022-06-17 21:16:41,486 # *** RIOT kernel panic:
2022-06-17 21:16:41,505 # FAILED ASSERTION.
2022-06-17 21:16:41,506 # 
2022-06-17 21:16:41,519 # *** halted.
```

```
 $ avr-addr2line -e bin/m1284p/hello-world.elf 0x104
/home/maribu/Repos/software/RIOT/examples/hello-world/main.c:29
```

Reply:
And I think with this it may actually work with AVRs with more than 128 KiB RAM (@kfessel may be able to confirm):

```diff
diff --git a/core/lib/assert.c b/core/lib/assert.c
index 3e3a3526f4..ba6296dec4 100644
--- a/core/lib/assert.c
+++ b/core/lib/assert.c
@@ -13,6 +13,8 @@
  * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
  */
 
+#include <inttypes.h>
+#include <stdint.h>
 #include <stdio.h>
 
 #include "assert.h"
@@ -27,7 +29,7 @@ __NORETURN void _assert_failure(const char *file, unsigned line)
 
 __NORETURN void _assert_panic(void)
 {
-    printf("%p\n", (void *)cpu_get_last_instruction());
+    printf("0x%" PRIx32 "\n", (uint32_t)cpu_get_last_instruction());
     core_panic(PANIC_ASSERT_FAIL, "FAILED ASSERTION.");
 }
 
diff --git a/cpu/avr8_common/include/cpu.h b/cpu/avr8_common/include/cpu.h
index c3eb747c37..ef26366d33 100644
--- a/cpu/avr8_common/include/cpu.h
+++ b/cpu/avr8_common/include/cpu.h
@@ -32,12 +32,13 @@
 #ifndef CPU_H
 #define CPU_H
 
-#include <stdio.h>
+#include <avr/interrupt.h>
 #include <stdint.h>
+#include <stdio.h>
 
-#include <avr/interrupt.h>
-#include "cpu_conf.h"
 #include "cpu_clock.h"
+#include "cpu_conf.h"
+#include "macros/xtstr.h"
 #include "sched.h"
 #include "thread.h"
 
@@ -166,24 +167,52 @@ void avr8_exit_isr(void);
  */
 void avr8_clk_init(void);
 
+/**
+ * @brief   True if the AVR8 has an EIND register (to address more than 128 KiB
+ *          program memory), otherwise false
+ */
+#define CHECK_EIND_REG    FLASHEND > 0x1ffff
+
+#ifndef __EIND__
+/**
+ * @brief   I/O address of the EIND reg (if existing)
+ */
+#define __EIND__                0x3C
+#endif
+
+
 /**
  * @brief   Get the last instruction's address
+ *
+ * @details There is absolutely no way to do this reliably on AVR. Instead, we
+ *          examine the stack frame and assume that the caller called this
+ *          function right away, so that the return address is still at the top
+ *          of the stack
  */
-static inline uintptr_t __attribute__((always_inline)) cpu_get_last_instruction(void)
+__attribute__((always_inline))
+static inline uint32_t cpu_get_last_instruction(void)
 {
-    uint8_t hi;
-    uint8_t lo;
-
-    __asm__ volatile (
-        "in %[hi], __SP_H__     \n\t"
-        "in %[lo], __SP_L__     \n\t"
-        : [hi] "=r"(hi),
-          [lo] "=r"(lo)
+    uint32_t result;
+    __asm__ volatile(
+        "ldi %D[dest], 0"                   "\n\t"
+#if (CHECK_EIND_REG)
+        "in %C[dest], " XTSTR(__EIND__)     "\n\t"
+#else
+        "ldi %C[dest], 0"                   "\n\t"
+#endif
+        "pop %B[dest]"                      "\n\t"
+        "pop %A[dest]"                      "\n\t"
+        "push %A[dest]"                     "\n\t"
+        "push %B[dest]"                     "\n\t"
+        : [dest]    "=r"(result)
         : /* no inputs */
-        : /* no clobbers */
+        : "memory"
     );
 
-    return hi << 8 | lo;
+    /* Result now contains instruction to return to, subtract one to get
+     * the instruction that called this function. Also multiply by two to get
+     * the byte position, rather than the (16 bit) instruction position */
+    return (result - 1) * 2;
 }
 
 /**
diff --git a/cpu/avr8_common/thread_arch.c b/cpu/avr8_common/thread_arch.c
index d07d6448b1..b3b0019155 100644
--- a/cpu/avr8_common/thread_arch.c
+++ b/cpu/avr8_common/thread_arch.c
@@ -34,8 +34,6 @@
 #include "board.h"
 #include "macros/xtstr.h"
 
-#define CHECK_EIND_REG          FLASHEND > 0x1ffff
-
 #if defined(DATAMEM_SIZE)
 #define CHECK_RAMPZ_REG         DATAMEM_SIZE > 0xffff || FLASHEND > 0xffff
 #define CHECK_RAMPDXY_REG       DATAMEM_SIZE > 0xffff
@@ -44,12 +42,6 @@
 #define CHECK_RAMPDXY_REG       0
 #endif
 
-#if (CHECK_EIND_REG)
-#ifndef __EIND__
-#define __EIND__                0x3C
-#endif
-#endif
-
 static void avr8_context_save(void);
 static void avr8_context_restore(void);
 static void avr8_enter_thread_mode(void);
```

Reply:
the result:
```
2022-06-17 21:54:16,687 # �
2022-06-17 21:54:16,688 # main(): This is RIOT! (Version: tool-test-fixed-1305-g54c0f0-HEAD)
2022-06-17 21:54:16,688 # Hello World!
2022-06-17 21:54:16,689 # 0xfffffffe
2022-06-17 21:54:16,713 # *** RIOT kernel panic:
2022-06-17 21:54:16,730 # FAILED ASSERTION.
2022-06-17 21:54:16,730 # 
2022-06-17 21:54:16,742 # *** halted.
2022-06-17 21:54:16,746 # 
2022-06-17 21:54:51,074 # Exiting Pyterm
```
seems wrong to me (did that on an arduino mega adk (similar to arduino-mega2560)

```
$ avr-objdump -CS bin/arduino-mega2560/core_lib/assert.o

Disassembly of section .text._assert_panic:

00000000 <_assert_panic>:
   0:   b0 e0           ldi     r27, 0x00       ; 0
   2:   ac b7           in      r26, 0x3c       ; 60
   4:   9f 91           pop     r25
   6:   8f 91           pop     r24
   8:   8f 93           push    r24
   a:   9f 93           push    r25
   c:   01 97           sbiw    r24, 0x01       ; 1
   e:   a1 09           sbc     r26, r1
  10:   b0 48           sbci    r27, 0x80       ; 128
  12:   88 0f           add     r24, r24
  14:   99 1f           adc     r25, r25
  16:   aa 1f           adc     r26, r26
  18:   bb 1f           adc     r27, r27
  1a:   bf 93           push    r27
  1c:   af 93           push    r26
  1e:   9f 93           push    r25
  20:   8f 93           push    r24
  22:   80 e0           ldi     r24, 0x00       ; 0
  24:   90 e0           ldi     r25, 0x00       ; 0
  26:   9f 93           push    r25
  28:   8f 93           push    r24
  2a:   0e 94 00 00     call    0       ; 0x0 <_assert_panic>
  2e:   60 e0           ldi     r22, 0x00       ; 0
  30:   70 e0           ldi     r23, 0x00       ; 0
  32:   83 e0           ldi     r24, 0x03       ; 3
  34:   0e 94 00 00     call    0       ; 0x0 <_assert_panic>
```
call adresses are 0 because its not linked

Reply:
```
__attribute__((always_inline))
static inline uint32_t cpu_get_last_instruction(void)
{
    uint32_t result;
    __asm__ volatile(
        "ldi %D[dest], 0"                   "\n\t"
#if (CHECK_EIND_REG)
        "pop %C[dest] "      "\n\t"
#else
        "ldi %C[dest], 0"                   "\n\t"
#endif
        "pop %B[dest]"                      "\n\t"
        "pop %A[dest]"                      "\n\t"
        "push %A[dest]"                     "\n\t"
        "push %B[dest]"                     "\n\t"
#if (CHECK_EIND_REG)
        "push %C[dest] "      "\n\t"
#endif
        : [dest]    "=r"(result)
        : /* no inputs */
        : "memory"
    );

    /* Result now contains instruction to return to, subtract one to get
     * the instruction that called this function. Also multiply by two to get
     * the byte position, rather than the (16 bit) instruction position */
    return (result - 1 ) * 2;
}
```

this works:

```

2022-06-17 22:44:26,718 # main(): This is RIOT! (Version: tool-test-fixed-1305-g54c0f0-HEAD)
2022-06-17 22:44:26,718 # Hello World!
2022-06-17 22:44:26,719 # 0x174
2022-06-17 22:44:26,726 # *** RIOT kernel panic:
2022-06-17 22:44:26,746 # FAILED ASSERTION.
2022-06-17 22:44:26,746 # 
2022-06-17 22:44:26,759 # *** halted.
2022-06-17 22:44:26,763 # 
2022-06-17 22:44:31,379 # Exiting Pyterm
```
```
$ avr-addr2line -e bin/arduino-mega2560/hello-world.elf 0x174
...... /examples/hello-world/main.c:27
```

Reply:
> seems wrong to me (did that on an arduino mega adk (similar to arduino-mega2560)

My fault, can you check this?

```diff
diff --git a/core/lib/assert.c b/core/lib/assert.c
index 3e3a3526f4..ba6296dec4 100644
--- a/core/lib/assert.c
+++ b/core/lib/assert.c
@@ -13,6 +13,8 @@
  * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
  */
 
+#include <inttypes.h>
+#include <stdint.h>
 #include <stdio.h>
 
 #include "assert.h"
@@ -27,7 +29,7 @@ __NORETURN void _assert_failure(const char *file, unsigned line)
 
 __NORETURN void _assert_panic(void)
 {
-    printf("%p\n", (void *)cpu_get_last_instruction());
+    printf("0x%" PRIx32 "\n", (uint32_t)cpu_get_last_instruction());
     core_panic(PANIC_ASSERT_FAIL, "FAILED ASSERTION.");
 }
 
diff --git a/cpu/avr8_common/include/cpu.h b/cpu/avr8_common/include/cpu.h
index c3eb747c37..c775305e91 100644
--- a/cpu/avr8_common/include/cpu.h
+++ b/cpu/avr8_common/include/cpu.h
@@ -32,12 +32,13 @@
 #ifndef CPU_H
 #define CPU_H
 
-#include <stdio.h>
+#include <avr/interrupt.h>
 #include <stdint.h>
+#include <stdio.h>
 
-#include <avr/interrupt.h>
-#include "cpu_conf.h"
 #include "cpu_clock.h"
+#include "cpu_conf.h"
+#include "macros/xtstr.h"
 #include "sched.h"
 #include "thread.h"
 
@@ -166,24 +167,45 @@ void avr8_exit_isr(void);
  */
 void avr8_clk_init(void);
 
+/**
+ * @brief   True if the AVR8 has an EIND register (to address more than 128 KiB
+ *          program memory), otherwise false
+ */
+#define CHECK_EIND_REG    FLASHEND > 0x1ffff
+
+
 /**
  * @brief   Get the last instruction's address
+ *
+ * @details There is absolutely no way to do this reliably on AVR. Instead, we
+ *          examine the stack frame and assume that the caller called this
+ *          function right away, so that the return address is still at the top
+ *          of the stack
  */
-static inline uintptr_t __attribute__((always_inline)) cpu_get_last_instruction(void)
+__attribute__((always_inline))
+static inline uint32_t cpu_get_last_instruction(void)
 {
-    uint8_t hi;
-    uint8_t lo;
-
-    __asm__ volatile (
-        "in %[hi], __SP_H__     \n\t"
-        "in %[lo], __SP_L__     \n\t"
-        : [hi] "=r"(hi),
-          [lo] "=r"(lo)
+    uint32_t result;
+    __asm__ volatile(
+        "ldi %D[dest], 0"                   "\n\t"
+#if (CHECK_EIND_REG)
+        "pop %c[dest]"                      "\n\t"
+#else
+        "ldi %C[dest], 0"                   "\n\t"
+#endif
+        "pop %B[dest]"                      "\n\t"
+        "pop %A[dest]"                      "\n\t"
+        "push %A[dest]"                     "\n\t"
+        "push %B[dest]"                     "\n\t"
+        : [dest]    "=r"(result)
         : /* no inputs */
-        : /* no clobbers */
+        : "memory"
     );
 
-    return hi << 8 | lo;
+    /* Result now contains instruction to return to, subtract one to get
+     * the instruction that called this function. Also multiply by two to get
+     * the byte position, rather than the (16 bit) instruction position */
+    return (result - 1) * 2;
 }
 
 /**
diff --git a/cpu/avr8_common/thread_arch.c b/cpu/avr8_common/thread_arch.c
index d07d6448b1..6f7bfd7ffd 100644
--- a/cpu/avr8_common/thread_arch.c
+++ b/cpu/avr8_common/thread_arch.c
@@ -34,8 +34,6 @@
 #include "board.h"
 #include "macros/xtstr.h"
 
-#define CHECK_EIND_REG          FLASHEND > 0x1ffff
-
 #if defined(DATAMEM_SIZE)
 #define CHECK_RAMPZ_REG         DATAMEM_SIZE > 0xffff || FLASHEND > 0xffff
 #define CHECK_RAMPDXY_REG       DATAMEM_SIZE > 0xffff
diff --git a/examples/hello-world/main.c b/examples/hello-world/main.c
index f51bf8c0a0..d8bbf3985b 100644
--- a/examples/hello-world/main.c
+++ b/examples/hello-world/main.c
@@ -19,12 +19,15 @@
  * @}
  */
 
+#include <assert.h>
 #include <stdio.h>
 
 int main(void)
 {
     puts("Hello World!");
 
+    assert(0);
+
     printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
     printf("This board features a(n) %s MCU.\n", RIOT_MCU);
 
```

Reply:
Looks like you just did. And it even works with the unmaintained AVR toolchain in the riotbuild

Reply:
also works with far away functions failing

```
void far_away(void) __attribute__((section("FAR_AWAY"),noinline));

int main(void)
{
    puts("Hello World!");
    far_away();
...
}


void far_away(void){
    assert(0);

}
```

```
$LINKFLAGS=-Wl,--section-start=FAR_AWAY=11000 BOARD=arduino-mega2560 make flash term
...
2022-06-17 23:34:53,687 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2022-06-17 23:34:54,693 # 

2022-06-17 23:34:54,694 # main(): This is RIOT! (Version: tool-test-fixed-1305-g54c0f0-HEAD)
2022-06-17 23:34:54,694 # Hello World!
2022-06-17 23:34:54,694 # 0x11002
2022-06-17 23:34:54,711 # *** RIOT kernel panic:
2022-06-17 23:34:54,732 # FAILED ASSERTION.
2022-06-17 23:34:54,732 # 
2022-06-17 23:34:54,744 # *** halted.
2022-06-17 23:34:54,748 #
```

Reply:
The patch I sent is missing the corresponding push for the CHECK_EIND case, @kfessel added it in his variant

Reply:
I mean this PR just moves code around - wouldn't this be more suited to a follow-up?

Reply:
Given that the solution is pasted here and tested, why delay?

Also keep in mind that the previous print it was obvious that the AVR code printed a stack pointer rather than a code address, as the function printed "stack pointer" right before the address. It was still not matching the API, but at least there was no silent mismatch between API an behavior. This would now be the case, which is a significant regression from master.

Finally, it exposed a bug in the API. The assumption that `uintptr_t` is suitable to hold a programm address doesn't hold. `uint32_t` would be suitable for all current targets, but 64 bit RISC-V support is on the way in. So we need a specific type for this along with `PRI<foo>` macros for printing...

Reply:
Would `void *` not work? (But then `uintptr_t` would be broken if it doesn't fit a whole pointer…)

Reply:
Note that `uintptr_t` is defined to be able to store any possible `void *` value without data loss. So `void *` cannot be larger than `uintptr_t`.

The issue is that AVR is a Harvard arch with a different address space for flash and SRAM. The data address space is 16 bit wide (well, there are magic RAM page select registers on some AVRs to extend to 24 bit address space, but let's ignore this here), but the program memory is 17 bit wide (as the 16 bit program counter drops the last bit, as instructions are always two byte wide) and can be extended on some AVRs to 25 bits (e.g. the ATmega2560 has 256 KiB of flash, that is more than 17 bit can address).

`void *` and `uintptr_t` are guaranteed to work fine for address space, but may not work when the program space is larger (such as e.g. on the ATmega2560).

Reply:
> Given that the solution is pasted here and tested, why delay?

Because one PR should be about exactly one issue. ;-)

Reply:
This PR is creating a new function (API) and removes a then obsolete old one (which we consider internal and -> not need to deprecate) even though this seems like copying it isn't.

If there are two commits in this it is:
- create (new API with correct implementation if possible or 0 implementation for fallback) (and adapt assert)
- and remove (old print function).
(but i do not think we need to split that part)

I think we should get the new interface right -> which means for AVR we cannot use `uintptr_t`. The type that most closely matches is `typedef void (*fptr_t)(void)` i think. Sadly even gcc does not handle that [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQATlIOAMngMmABy7gBGmMTepAAOqAqEtgzObh7RcQk2AgFBoSwRUT6WmNZJQgRMxAQp7p5FmFZZDOWVBDkh4ZHRChVVNWlFvW2BHfldXgCUFqiuxMjsHAD0iwDUACoAnjGYKxszxCtoWCsIkZikKyQrtKhM6CuGK5iqrDH0AHQApBoAggRbmCwVBWADdUHh7hAAFRUGIEYjGAgTCBgiETT4AZgAQt8fqj7mwWMZkDcwhBAgRSO9qejsbj6b8KSsFABHVyVTDkwQrBjuCYrT4AdhxvxWYpWsPhiJWBAFGIAIitqHCEUiNKouBotRpMSKfuKVoTiaSIFxaXqDQQIObcQbiJgCLMGCsdXTfkL5RwprROABWXieDhaUioTiOZn7eYCgBMGJ4pAImi9UwA1iAMRj3hnszncwA2fScSQBpMhzi8BQgDQJpNTOCwJBoFgxOiRciUJst%2BhRUQg4jALia6s0WgESKViBhUthQKVDaceNNtiCADyDFo86DvCwLEMwHEW9I%2BHtpRBmErh%2BeJVcY4XvApDVLtDwYWIc%2BcWFL8LwLDvUyoBjAAoABqeCYAA7iu2yBvG/CCCIYjsFIMiCIoKjqIeuhcPoe4gKYxL6C%2BlaQFMqBwkkF4ALQsCwyCuJi8pMAQbDAEw0a%2BnmwbFKUdgQA4/SeNh/gjHkBR6BkiQCAJ4nxJJDDtKJXTYdxTQtH0Li1HoKllEMCmdFEylDNJhmtHpYwGVMCiRkhCb2gsPDen6JaHqGHA/MBABKKzAMgyArIO7xcO8GhKo4Fy4IQVyfLG2ErM4zatgc0VxhMvCJluEypumma5rl2YFj6HDFqQgbBq5FZVjWGWkPWiAoKgCXdu2ECdolKBMH2A5DnwdBjsQE5ToeM7MMQm6Lg1y4EGuG6ljue4HsGx4lDYZ4XsGV50be3D3oIj6Hs%2Br7vhgCzBt%2Bv7bf%2BgEgWBkHQXeKHCKI4jIXB8hKGopa6NGOFGHhZiEWExHWiG5ECFRNF0QxTEsUwGLRgAHDEFYNMtST2AwTgaWkQkY2ZYnYRJTTGbEslNHjSkWCjPHNEZWOCZTjQ6aZIn6VptOpPTPTM7krNmtMszzHo8KYPZXqFhw/olaWrnuV5Pl%2BQFQUhRAYUrBFRBJTFFzxV2kQxnDqVVVomWkGmGZZnluXi8VpW8OVFiVelxuORw0bOWV5ZG8mpBnv1aOSEAA%3D%3D)

GCCs type for goto pointers  is/will be `void *`.  (https://gcc.gnu.org/gcc-12/porting_to.html)

Reply:
according to https://gcc.gnu.org/onlinedocs/gcc-10.2.0/gcc/AVR-Options.html there is a 
`__AVR_3_BYTE_PC__` or a `__AVR_2_BYTE_PC__` define maybe we should use that instead of `if (CHECK_EIND_REG)`

Reply:
Split out the fix as requested: https://github.com/RIOT-OS/RIOT/pull/18234

Let's have this fixed prior to merging this PR.